### PR TITLE
NegoEx implementation

### PR DESCRIPTION
**Pull request by lhoward**
Relicense & rebase of NegoEx implementation I wrote some years ago for interoperability with Windows. At this stage have only checked it compiles, the last time I tested it worked was in 2015.

### CODE REVIEW

**Comment by lhoward** [comment]
build failing as make depend needs to be run

**Comment by greghudson** [comment]
Thanks for this.  I'll want to review the code carefully against the spec, which will be time-consuming, and I have some things to finish up, so this will sit here for a while.

You can add an empty lib/gssapi/negoex/deps file if you want to clear up the build error.

I would be interested to hear about specific interoperability stories for NegoEx.  I have a general idea of what it is and how it improves on SPNEGO, but I don't know if specific Microsoft products require NegoEx to interop.

**Comment by lhoward** [comment]
MS uses it for PKU2U. There's a general issue from Windows 8 onwards where new Windows mechanisms that do not support NegoEx have to be preferenced ahead of NTLM in order to be negotiable with SPNEGO (which isn't always desirable). There are some ugly parts of the code such as writing packed structures out directly on the wire. I'll regenerate the deps this morning.

**Comment by lhoward** [comment]
(From Windows 8 onwards mechanisms can only advertise themselves as supporting NegoEx or SPNEGO. My understanding is that all newer Microsoft mechanisms from and including PKU2U require NegoEx to be negotiable.)

**Comment by lhoward** [comment]
No rush too, I wouldn't integrate until I've had a chance to retest against the EAP SSP on Windows, which may be challenging as I'm not sure if my Windows development machine boots anymore.

**Comment by lhoward** [comment]
In terms of features, a nice attribute of NegoEx is the ability to exchange some credential-related metadata before committing to a mechanism.

**Comment by lhoward** [comment]
There’s a fair bit of abstraction in the code as I had planned to port to Heimdal someday. And yes, the fprintfs need to replaced!

**Comment by lhoward** [comment]
Cc @nicowilliams

**Comment by lhoward** [comment]
MIT to MIT is working.

**Comment by lhoward** [comment]
It's failing to build right now because of unaligned access, which is something I need to fix otherwise this won't be very portable.

**Comment by greghudson** [comment]
Be aware of k5-input.h if you need to write unmarshalling code.  It and k5-buf.h are roughly the MIT krb5 analog to Heimdal's krb5_storage.  (MIT krb5 doesn't have 16/32-bit integer helpers in k5-buf.h at this time, but it may be about time to add them between ccmarshal.c, cc_kcm.c, plugins/kdb/lmdb/marshal.c, and this code.)

**Comment by lhoward** [comment]
Ah, that looks good. I want to keep things sufficiently abstracted so I can port it to Heimdal later with ease, too. It'll take me a little while to untangle things...

**Comment by lhoward** [comment]
BTW @greghudson, `get_available_mechs()` can be simplified by including `GSS_C_MA_MECH_NEGO` in the list of excepted attributes, then you could just set `rmechs` directly. (OK, it's true that that will exclude *all* negotiation mechanisms but realistically I think that is *probably* fine, you may have a different opinion though. I need separate logic in any case to avoid NegoEx being negotiated by SPNEGO unless there are actually NegoEx-supporting mechanisms installed.)

**Comment by lhoward** [comment]
Updates:

* have fixed some alignment issues (without writing correct marshallers, noted)
* interop tested with Windows 10 initiator
* fixed mechanism error token status propagation in NegoEx and SPNEGO
* hid NegoEx from gss_indicate_mechs()
* various other fixes

**Comment by simo5** [comment]
I like this, any chance we can get some (interop even?) tests ?

**Comment by lhoward** [comment]
Hey @simo5, I have been testing with mech_eap (with a small modification to set `GSS_C_MA_NEGOEX_ONLY`, although it will work fine without), with gss-sample both on MIT and Windows. I will find out how to get you a copy of the EAP SSP for Windows as that's probably the easiest thing to test interop as we don't have a PKU2U implementation for MIT.

**Comment by lhoward** [comment]
As for automated testing, we'd need a dummy mechanism or perhaps a way to make krb5 work under NegoEx just for testing. That's more than I have time to do right now though.

**Comment by lhoward** [comment]
[MS-SPNG] also has some hex dumps if that's useful.

**Comment by lhoward** [comment]
Links:

https://www.ietf.org/proceedings/72/slides/kitten-2.pdf

https://msdn.microsoft.com/en-us/library/ff468736.aspx

**Comment by cryptomilk** [comment]
Luke, do you also have wireshark support with decryption for it?

**Comment by lhoward** [comment]
Andreas, nice to hear from you. I don’t although I did see something on the mailing lists about one, not sure if it was ever implemented.

**Comment by lhoward** [comment]
Also, there’s no encryption in the protocol, only a checksum. The rest is clear text although of course the protocols it negotiates might use encryption in setting up their security contexts.

**Comment by lhoward** [comment]
> Thanks for dredging this back up!
> 
> At a high level, I'd like to see both in-tree testing (I recognize this is nontrivial) and successful operation against Microsoft before we consider inclusion. And of course, you need to pass CI :)

Got it. I'm back at my other job at the moment but I will get to it over the coming months. It does interoperate with Windows but this may be difficult for you to test (but I can get you access to the EAP SSP if you contact me privately).

File: src/lib/gssapi/mechglue/g_negoex.c [hunk 1]
```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright 2011 by the Massachusetts Institute of Technology.
```
[the remark this answers is not in the document]
Comment (frozencemetery) [reply src/lib/gssapi/mechglue/g_negoex.c]: Just double-checking.  The others say 2011-2018 PADL, while this is the only file that says 2011 MIT.  If you say it's fine then it's of course fine :)

**Comment by greghudson** [comment]
For in-tree testing we need to be able to mock GSS mechanisms.  (We also need that ability to properly test a lot of SPNEGO code.)  I can think of three basic directions for making that possible:

* Make a private C interface that allows us to register a mechanism in the mech table using a struct gss_config.

* Make a private C interface that allows us to register a DLL in the mech table, as if it were listed in /etc/gss/mechs.  We would build the mock GSS module the way we build other test programs.

* Make the mechglue look at an environment variable or at krb5.conf.

I will probably try the first option and then fall back to the second one if the first one gets messy for some reason, but I'm open to opinions.

**Comment by lhoward** [comment]
If you only need to test on platforms that support dynamic loading, then you could have an environment variable that points to where to read /etc/gss/mech from, and set that in the tests.

**Comment by frozencemetery** [comment]
Fixing the /etc/gss/mech.d interference would be nice.  Downstream we have to be careful not to have NTLM/gssproxy mechs lying around during testing.

How do we handle KRB5_CONFIG with setuid programs (i.e., ksu) today?  In general `secure_getenv()` seems like a good idea (we use it in gssproxy), but I suppose it is a GNU extension.  Heimdal's support pulls out a separate setuid check that we'd probably want to inline, but otherwise doesn't look unreasonable.

A new environment variable probably isn't a good idea for the reasons you mentioned above, but a (possibly private) configuration in krb5.conf might not be a bad idea.

**Comment by greghudson** [comment]
For programs using libkrb5, we have krb5_init_secure_context(), which causes environment variables to be ignored by libkrb5.  We don't have a story for setuid GSS apps.  Adding portability code for secure_getenv() is on the list.

Historically, environment pollution has not been limited to setuid processes.  Some of the earliest security holes were in remote login daemons which would accept environment variables from clients.  (Even if those didn't affect the authentication process, they could be used to escape restricted shells.)  At the time the fix was to sanitize the environments of certain variables, which wasn't very scalable.  It's possible that this kind of bug is no longer an issue.

**Comment by frozencemetery** [comment]
Hi @lhoward, have you had time to work on this recently?

**Comment by greghudson** [comment]
I'm expecting to get to this and am not waiting on Luke, just my own time availability.  I've implemented secure_getenv(), so I can add an environment variable for /etc/gss/mech and one or more test mechs for testing.

**Comment by lhoward** [comment]
Thanks, I can take a look at the conflicts against master but I don't have time to write a new mech for testing. I can test whatever Greg integrates against Windows with GSS EAP though, to verify it still works.

**Comment by greghudson** [comment]
Just doing a very preliminary inventory of the PR changes outside of the negoex directory itself:

* There is some stuff in here for GSS_S_PROMPTING_NEEDED, which is not mentioned in draft-zhu-negoex-04.  I'm guessing this has something to do with Moonshot but is not strictly related to NegoEx.

* gss_indicate_mechs_by_attrs() is changed to query the unfiltered list, not the list with deprecated mechanisms pruned.  That behavior change needs justification.

* There's an apparent bugfix to gss_set_context_option() to set the loopback field when a union context is allocated.  That needs to be a separate commit, with an analysis of the bug's impact.  (Are there existing cases for creating contexts with gss_set_context_option() and are they always broken? etc.)

* I wonder if NegoEx needs to be visible as a mech to the mech switch, or if it could be implemented internally to SPNEGO.  Right now NegoEx is hidden from some, but not all, mechglue operations, and SPNEGO presumably gets to negoex_gss_init_sec_context(), negoex_gss_acquire_cred() and negoex_gss_accept_sec_context() via the mechglue.  But I'm not sure that's necessary.  More research necessary.  (The three new entry points for NegoEx-only mechanisms do need to be part of the mech switch, of course.  I wonder if they should have gsspi_ prefixes?  That change would probably be obnoxious to Moonshot since Heimdal merged this months ago with gss_ prefixes.)

**Comment by lhoward** [comment]
Thank you for taking the time to look at this @greghudson.

> * There is some stuff in here for GSS_S_PROMPTING_NEEDED, which is not mentioned in draft-zhu-negoex-04.  I'm guessing this has something to do with Moonshot but is not strictly related to NegoEx.

Correct, I actually added it for CredUI, a credential acquisition provider interface inspired by the Windows interface of the same name. We can factor that out if you like.

> * gss_indicate_mechs_by_attrs() is changed to query the unfiltered list, not the list with deprecated mechanisms pruned.  That behavior change needs justification.

Where is this exactly? In `prune_to_indicated_mechs()`?

> * There's an apparent bugfix to gss_set_sec_context_option() to set the loopback field when a union context is allocated.  That needs to be a separate commit, with an analysis of the bug's impact.  (Are there existing cases for creating contexts with gss_set_sec_context_option() and are they always broken? etc.)  [Update: it doesn't look like there are any sec context options in the in-tree mechs yet.]

I guess that's probably just something I noticed along the way. I think.

> * I wonder if NegoEx needs to be visible as a mech to the mech switch, or if it could be implemented internally to SPNEGO.  Right now NegoEx is hidden from some, but not all, mechglue operations, and SPNEGO presumably gets to negoex_gss_init_sec_context(), negoex_gss_acquire_cred() and negoex_gss_accept_sec_context() via the mechglue.  But I'm not sure that's necessary.  More research necessary.  (The three new entry points for NegoEx-only mechanisms do need to be part of the mech switch, of course.  I wonder if they should have gsspi_ prefixes?  That change would probably be obnoxious to Moonshot since Heimdal merged this months ago with gss_ prefixes.)

The new NegoEx SPIs aren't in Heimdal although they are in Moonshot. But there is no shipping mechglue that uses them so this could be changed if you like. In some respects it would be cleaner to implement entirely within SPNEGO (because you would avoid the possibility of NegoEx ever being invoked directly), on the other hand I imagine it would be much more intrusive to SPNEGO. I felt the latter was the lesser of the two evils.

**Comment by greghudson** [comment]
>> gss_indicate_mechs_by_attrs() is changed to query the unfiltered list, not the list with deprecated
>> mechanisms pruned. That behavior change needs justification.
> Where is this exactly? In prune_to_indicated_mechs()?

In g_initialize.c, is_deprecated() is changed to is_indicated() and also filters out the new NOT_INDICATED attribute.  But also, gssint_indicate_mechs() is added as an unfiltered gss_indicate_mechs(), and gss_indicate_mechs_by_attrs() is changed to call it.

> The new NegoEx SPIs aren't in Heimdal although they are in Moonshot.

Huh, I thought Heimdal had merged this, but I must be confusing it with the KEYRING ccache work.

The gssspi_ prefix is traditionally used for contracts between the mechglue and mech implementations, like gssspi_import_name_by_mech().  (It's also used for acquire_cred_with_password for some reason.)  These entry points are a little different because they are a private contract between a specific negotiation mechanism and mech implementations.  draft-zhu-negoex uses a gss_ prefix.  Since it isn't totally clear that gssspi_ is appropriate, perhaps we should leave the name alone.

**Comment by lhoward** [comment]
> > > gss_indicate_mechs_by_attrs() is changed to query the unfiltered list, not the list with deprecated
> > > mechanisms pruned. That behavior change needs justification.
> > > Where is this exactly? In prune_to_indicated_mechs()?
> 
> In g_initialize.c, is_deprecated() is changed to is_indicated() and also filters out the new NOT_INDICATED attribute. But also, gssint_indicate_mechs() is added as an unfiltered gss_indicate_mechs(), and gss_indicate_mechs_by_attrs() is changed to call it.

Hmm, that seems weird, it should return not indicated (but not filter out deprecated), right? Sounds like a bug.

> The gssspi_ prefix is traditionally used for contracts between the mechglue and mech implementations, like gssspi_import_name_by_mech(). (It's also used for acquire_cred_with_password for some reason.) These entry points are a little different because they are a private contract between a specific negotiation mechanism and mech implementations. draft-zhu-negoex uses a gss_ prefix. Since it isn't totally clear that gssspi_ is appropriate, perhaps we should leave the name alone.

Practically I think either is fine. I used gss_ because that was what was in the draft, but gssspi_ (which I think I added!) is probably technically more correct (but couldn't have expected Larry to know about that).

**Comment by greghudson** [comment]
> Correct, I actually added it for CredUI, a credential acquisition provider interface inspired by the Windows interface of the same name. We can factor that out if you like.

It shouldn't be part of this PR.  In addition to being out of scope, it allocates a GSS supplementary status bit without IETF action.

> In some respects it would be cleaner to implement entirely within SPNEGO (because you would avoid the possibility of NegoEx ever being invoked directly)

I am particularly concerned about the path to NegoEx through gss_accept_sec_context() with the default acceptor cred (or with an acceptor cred acquired with no specified mech set).  That's not just an unwanted API surface; it's an unwanted protocol surface.

**Comment by lhoward** [comment]
How about setting a piece of thread specific data indicating whether we are inside SPNEGO or not?

**Comment by greghudson** [comment]
I don't appear to have permission to push changes (checkbox on the right side at the bottom).

**Comment by lhoward** [comment]
Allow edits from maintainers is checked, is this the one?

**Comment by greghudson** [comment]
> Allow edits from maintainers is checked, is this the one?

That's the one.  I don't know why I can't push to the branch.  I had this problem with PR #579, which also had a `/` in the branch name, but I think in that case it was my fault (I hadn't noticed the "asn/" prefix).

> How about setting a piece of thread specific data indicating whether we are inside SPNEGO or not?

That will work in a pinch.  I will try to investigate what it would mean to fold the negoex code into SPNEGO, and if it seems to messy we can fall back to thread-specific data or something.

**Comment by lhoward** [comment]
If there’s an elegant way by all means, but if it involves reinventing the mechglue then I reckon TSD is the way to go.

**Comment by greghudson** [comment]
This was moved to PR #980 for permissions-related reasons, and has been merged.

**Pull request closed without merging**